blaze-server/src/main/scala/org/http4s/server/blaze/Http1ServerStage.scala

```diff
@@ -332,13 +336,12 @@ private[blaze] class Http1ServerStage[F[_]](
     renderResponse(req, resp, bodyCleanup) // will terminate the connection due to connection: close header
   }
 
-  private[this] val raceTimeout: Request[F] => F[Response[F]] =
+  private[this] def raceTimeout(req: Request[F]): F[(Response[F], F[Unit])] =
     responseHeaderTimeout match {
       case finite: FiniteDuration =>
-        val timeoutResponse = timer.sleep(finite).as(Response.timeout[F])
-        req =>
-          F.race(runApp(req), timeoutResponse).map(_.merge)
+        val timeoutResponse = timer.sleep(finite).as((Response.timeout[F], F.unit))
+        F.race(run(req).allocated, timeoutResponse).map(_.merge)
```

Comment:
`race` doesn't guarantee that the finalizers on each of the parts "see" them being cancelled. E.g. in this example the resource is not released, even though `race` waits for allocation to finish before continuing:
```scala
val slowRes = Resource.make(IO {
  Thread.sleep(1000)
  println("Acquired")
})(_ => IO { println("Released") })

IO.race(slowRes.allocated, IO { Thread.sleep(500) })
  .flatMap(_.left.toOption.map(_._2).sequence_)
```

Reply:
This gets curiouser: https://github.com/typelevel/cats-effect/issues/579.

Reply:
Here's a stab at solving this issue from another angle, without diving into the `allocated`/`continual`/`uncancelable` rabbit hole:
```scala
def raceResource[F[_]: Concurrent, A](
  res: Resource[F, A],
  timeout: F[A]
): Resource[F, A] = for {
  earlyRelease <- Resource.liftF(Deferred[F, Unit])
  acquired <- Resource.liftF(Deferred[F, Either[Throwable, A]])

  _ <- res.attempt.use(
    acquired.complete(_) *> earlyRelease.get
  ).background

  a <- Resource.liftF(
    timeout.attempt.race(acquired.get)
      .flatTap { result =>
        // If the timeout has won, we can release the resource early.
        earlyRelease.complete(()).whenA(result.isLeft)
      }.map(_.merge)
  )
} yield a
```

This relies only on high-level combinators, and it shouldn't mask cancellation anywhere (there are no `Resource.make` or `uncancelable` calls). I haven't tested it, though.

Reply:
It has been a while since I looked at this, and I can't remember whether we can add this without breaking binary compatibility.  Not as the draft currently stands, but if we are more careful in the blaze builder and grant MiMa exceptions on private API, then probably yes?

If so, this would be neat to try to get into the 0.21 series.  If not, I like your thinking above, but we'll soon have to port it to Cats Effect 3.

Reply:
> without diving into the allocated/continual/uncancelable rabbit hole

The other thing to mention is that CE3 story on this stuff is a million times better, in general (although I haven't tried this specific example, so you never know)

Reply:
What you wrote above is not strictly speaking a `race` I don't think, a `Resource` composed of two slow `make` calls which times out during the first `make` won't be interrupted until _both_ complete

Reply:
> The other thing to mention is that CE3 story on this stuff is a million times better, in general (although I haven't tried this specific example, so you never know)

Yes, probably (I haven't really touched it yet), but I'd still rather deal with `Bracket`-like behavior over `allocated`.

> What you wrote above is not strictly speaking a race I don't think

I don't know what to call it. It encapsulates a pattern in this draft, where resource acquisition races against some timeout.

> a Resource composed of two slow make calls which times out during the first make won't be interrupted until both complete

Wait, so a `Resource` can't be interrupted between two `Allocate`s? That's surprising.

Still, I'm not sure how that applies to the code I wrote above. Unless you're saying that `res.use {... }` can't be cancelled during acquisition.

Reply:
> a Resource composed of two slow make calls which times out during the first make won't be interrupted until both complete

I've made a [scastie](https://scastie.scala-lang.org/AFuMkDP5RJ2rtidKV2ZIRw) to play around. I think I now see what you mean: `earlyRelease` does not cancel the *acquisition* of `res`: note that `bar` is still acquired even after the timeout.

Here's a revised `raceResource` [without this issue](https://scastie.scala-lang.org/srfhyAb6QXKvOehzG5IwRQ) (I've changed the `res.attempt` part):
```scala
def raceResource[F[_], A](
  res: Resource[F, A],
  timeout: F[A]
)(implicit F: Concurrent[F]): Resource[F, A] = for {
  earlyRelease <- Resource.liftF(Deferred[F, Unit])
  acquired <- Resource.liftF(Deferred[F, Either[Throwable, A]])

  _ <- res.attempt.use(
    acquired.complete(_) <* F.never
  ).race(earlyRelease.get)
    .background

  a <- Resource.liftF(
    timeout.attempt.race(acquired.get)
      .flatTap { result =>
        // If the timeout has won, we can release the resource early.
        earlyRelease.complete(()).whenA(result.isLeft)
      }.map(_.merge)
      .flatMap(_.liftTo[F])
  )
} yield a
```

Reply:
With CE3 we can just use `Resource.race` from the library instead of all this.